DNN Platform/Library/Services/Installer/Installers/ScriptInstaller.cs

```diff
@@ -187,6 +202,19 @@ public override void Install()
                 {
                     bSuccess = this.InstallScriptFile(this.UpgradeScript);
                     installedVersion = this.UpgradeScript.Version;
+                }
+
+                // Next process SchemaScripts - these script always runs if present

```

Comment:
I would think we would want these to run before versioned scripts.  But, I guess, regardless, I'm having a hard time seeing how you'd handle data migration if you're using these.  I guess you'd use schema scripts for views, functions, and procedures, but stick to versioned files for tables?  So, if that's the case, running schema scripts last makes sense to me.  (So, I don't think there's anything to change, but leaving this comment in case anyone wants to comment)

Reply:
I don't have experience using the install or upgrade scripts, so I may be missing some context on how these actually get used.  That said, it seems like, regardless of whether the schema scripts run before or after the versioned scripts, that the upgrade script is designed to run last.  Again, I may be missing something, but I would think we'd want the upgrade script to run after these schema scripts.

Reply:
I typically used the versioned scripts myself thus far, but I believe that the "install" script concept would work similar to that of how it works inside of DNN platform as a whole.

Install - This is used to setup the "base" of the installation.  It is processed first, and once completed the version that you provide with it is used to determine which additional scripts (if included) are used for processing.

Versioned Scripts - These then execute, as we expect, version-by-version starting with the next version after the "install" version.

Schema - This, as I see it would be a "replacement" of sorts for a lot of what might get done in the current Install or Versioned scripts.  You can do whatever you want here, but you have to realize that this is executed on every (initial) installation of your extension.

Now, what is actually done in here could be interesting....

For example, if using EF inside of a module, you could use an Idempotent script flag to generate your model and use the single Schema file.  This would work for all changes, including incremental schema changes.

Another example would be more limited where you put all SP's etc in there that you drop/create every time.  But table migrations would need to be done otherwise.

My biggest thoughts are that this is good, but I'm worried that the current naming conventions and black magic automatic assumptions might make this more complicated.

Reply:
I use install scripts from time to time. But mostly the versioned scripts, like @mitchelsellers .

About the purpose: as I point out, I see them being used in parts of the schema that can/need to be recreated. Take views: if you've changed a table "Foo" and added a column "Bar" and you have a view that includes "SELECT * FROM Foo" then Bar will not be present unless you recreate the view. 

I regularly deal with this as my pattern is to have a view in 99% of the tables I create and in code I use an object that inherits from the base object of the table. That way I can do CRUD operations easy where I expect the view for reading and the table (base object) for writing. Anyway, it only works if I meticulously keep track of the views as well.

So in terms of run order: I'd want the versioned scripts to run first. They do the most invasive stuff like work on data and change tables. I'm not sure about the upgrade scripts as I don't use those but I'd leave them as is just after. The schema scripts are really just a layer on top of the data that my code interacts with. Hence I'd keep it until last.

Reply:
Versionsed scripts need to be kept for table modifications (schema updates), everything else can be dropped and recreated using the other scripts. This way, they must be modified once only (no complex creation of rollup scripts anymore) and views and sprocs may be edited "in place" in these files - making file versions and upgrade diffs much easier to follow for devs.

Reply:
Versionsed scripts need to be kept for table modifications (schema updates), everything else can be dropped and recreated using the other scripts. This way, they must be modified once only (no complex creation of rollup scripts anymore) and views and sprocs may be edited "in place" in these files - making file versions and upgrade diffs much easier to folllow for devs.

Reply:
This will also allow using advanced features like indexed views, which may support performance dramatically.